node/pod.go

```diff
@@ -213,8 +191,12 @@ func (pc *PodController) updatePodStatus(ctx context.Context, podFromKubernetes
 	}
 	kPod := obj.(*knownPod)
 	kPod.Lock()
-	podFromProvider := kPod.lastPodStatusReceivedFromProvider
+	podFromProvider := kPod.lastPodStatusReceivedFromProvider.DeepCopy()
 	kPod.Unlock()
+	// We need to do this because the other parts of the pod can be updated elsewhere. Since we're only updating
+	// the pod status, and we should be the sole writers of the pod status, we can blind overwrite it. Therefore
+	// we need to copy the pod and set ResourceVersion to 0.
+	podFromProvider.ResourceVersion = "0"
```

Comment:
I don't understand this and don't see an example of this in the kubelet code.
Does the same issue exist in Kubelet?

Reply:
So, the way the kubelet does it is by fetching the current pod (status) from api server, and then does a client-side merge, before doing a server side patch. Code is attached:

```
// syncPod syncs the given status with the API server. The caller must not hold the lock.
func (m *manager) syncPod(uid types.UID, status versionedPodStatus) {
        if !m.needsUpdate(uid, status) {
                klog.V(1).Infof("Status for pod %q is up-to-date; skipping", uid)
                return
        }

        // TODO: make me easier to express from client code
        pod, err := m.kubeClient.CoreV1().Pods(status.podNamespace).Get(status.podName, metav1.GetOptions{})
        if errors.IsNotFound(err) {
                klog.V(3).Infof("Pod %q (%s) does not exist on the server", status.podName, uid)
                // If the Pod is deleted the status will be cleared in
                // RemoveOrphanedStatuses, so we just ignore the update here.
                return
        }
        if err != nil {
                klog.Warningf("Failed to get status for pod %q: %v", format.PodDesc(status.podName, status.podNamespace, uid), err)
                return
        }

        translatedUID := m.podManager.TranslatePodUID(pod.UID)
        // Type convert original uid just for the purpose of comparison.
        if len(translatedUID) > 0 && translatedUID != kubetypes.ResolvedPodUID(uid) {
                klog.V(2).Infof("Pod %q was deleted and then recreated, skipping status update; old UID %q, new UID %q", format.Pod(pod), uid, translatedUID)
                m.deletePodStatus(uid)
                return
        }

        oldStatus := pod.Status.DeepCopy()
        newPod, patchBytes, err := statusutil.PatchPodStatus(m.kubeClient, pod.Namespace, pod.Name, *oldStatus, mergePodStatus(*oldStatus, status.status))
        klog.V(3).Infof("Patch status for pod %q with %q", format.Pod(pod), patchBytes)
        if err != nil {
                klog.Warningf("Failed to update status for pod %q: %v", format.Pod(pod), err)
                return
        }
        pod = newPod

        klog.V(3).Infof("Status for pod %q updated successfully: (%d, %+v)", format.Pod(pod), status.version, status.status)
        m.apiStatusVersions[kubetypes.MirrorPodUID(pod.UID)] = status.version

        // We don't handle graceful deletion of mirror pods.
        if m.canBeDeleted(pod, status.status) {
                deleteOptions := metav1.NewDeleteOptions(0)
                // Use the pod UID as the precondition for deletion to prevent deleting a newly created pod with the same name and namespace.
                deleteOptions.Preconditions = metav1.NewUIDPreconditions(string(pod.UID))
                err = m.kubeClient.CoreV1().Pods(pod.Namespace).Delete(pod.Name, deleteOptions)
                if err != nil {
                        klog.Warningf("Failed to delete status for pod %q: %v", format.Pod(pod), err)
                        return
                }
                klog.V(3).Infof("Pod %q fully terminated and removed from etcd", format.Pod(pod))
                m.deletePodStatus(uid)
        }
}

```

Reply:
ping @pires

Reply:
The “0” thing is a “regression”. In 1.0, we used to do it here:
https://github.com/virtual-kubelet/virtual-kubelet/blob/release-1.0/node/pod.go#L114